src/ppom-storage.ts

```diff
@@ -0,0 +1,206 @@
+ * It provides the following functionalities:
+ * 1. Sync the metadata with the version info from the cdn
+ * 2. Read a file from the local storage
+ * 3. Write a file to the local storage
+ *
+ * It also validates the checksum of the file when reading and writing in order to
+ * detect corrupted files or files that are not up to date
+ */
+export class PPOMStorage {
+  readonly #storageBackend: StorageBackend;
+
+  readonly #readMetadata: () => PPOMFileMetadata;
+
+  readonly #writeMetadata: (metadata: PPOMFileMetadata) => void;
+
+  /**
+   * Creates a PPOMStorage instance.
+   *
+   * @param options - The options passed to the function.
+   * @param options.storageBackend - The storage backend to use for the local storage.
+   * @param options.readMetadata - A function to read the metadata from the local storage.
+   * @param options.writeMetadata - A function to write the metadata to the local storage.
+   */
+  constructor({
+    storageBackend,
+    readMetadata,
+    writeMetadata,
+  }: {
+    storageBackend: StorageBackend;
+    readMetadata: () => PPOMFileMetadata;
+    writeMetadata: (metadata: PPOMFileMetadata) => void;
+  }) {
+    this.#storageBackend = storageBackend;
+    this.#readMetadata = readMetadata;
+    this.#writeMetadata = writeMetadata;
+  }
+
+  /**
+   * Sync the metadata with the version info from the cdn.
+   * 1. Remove the files that are not readable (e.g. corrupted or deleted).
+   * 2. Remove the files that are not in the cdn anymore.
+   * 3. Remove the files that are not up to date in the cdn.
+   * 4. Remove the files that are not in the local storage from the metadata.
+   * 5. Delete the files that are not in the metadata from the local storage.
+   *
+   * @param versionInfo - Version information of metadata files.
+   */
+  async syncMetadata(versionInfo: PPOMFileMetadata): Promise<PPOMFileMetadata> {
+    const metadata = this.#readMetadata();
+    const syncedMetadata = [];
+
+    for (const fileMetadata of metadata) {
+      // check if the file is readable (e.g. corrupted or deleted)
+      try {
+        await this.readFile(fileMetadata.name, fileMetadata.chainId);
+      } catch (exp: any) {
+        continue;
+      }
+
+      // check if the file exits and up to date in the storage
+      if (
+        !versionInfo.find(
+          (file) =>
+            file.name === fileMetadata.name &&
+            file.chainId === fileMetadata.chainId &&
+            file.version === fileMetadata.version &&
+            file.checksum === fileMetadata.checksum,
+        )
+      ) {
+        continue;
+      }
+
+      syncedMetadata.push(fileMetadata);
+    }
+
+    const filesInDB = await this.#storageBackend.dir();
+    for (const { name, chainId } of filesInDB) {
+      if (
+        !syncedMetadata.find(
+          (file) => file.name === name && file.chainId === chainId,
+        )
+      ) {
+        await this.#storageBackend.delete({ name, chainId });
+      }
+    }
```

Comment:
maybe we can move this into a separate function. something like
 
`this.#clearUnsyncedMetadata(syncedMetadata)`

In my head, one of the aims is to shorten functions when we can and have them read more like a book. I think this helps with readability and supports our pursuit of SRP functions.

Reply:
The main function is only 40 lines, if we divide it into 3 - it may not be very understandable.